### ISSUE TITLE
Restrict primTrustMe to only reduce for canonical terms

### ISSUE THREAD

**Issue by UlfNorell** [agda/agda#3092]
Current `primTrustMe` reduces to `refl` when the normal forms of its arguments are syntactically equal. For instance,
```agda
is-refl : (a : Nat) → a ≡ a
is-refl a = primTrustMe
-- C-c C-n is-refl
-- λ a → refl
```
While this might be nice, it's not a conservative extension of the type theory. It's also not really needed for the typical applications of `primTrustMe`:
- implementing decidable equality for builtin types
- erasing big proof terms

One can even argue that having `primTrustMe` reduce on neutral arguments is a bit strange in both these cases.

Therefore, I propose to change the implementation of `primTrustMe` so that it only reduces when the arguments are canonical first-order terms.

As an added bonus this will make #1433 obsolete.

**Comment by andreasabel** [comment]
> canonical first-order terms

But `a : Nat` is a canonical first-order term! Or what do you mean by that?

**Comment by UlfNorell** [comment]
A closed normal form. Basically it should reduce in those cases where a hand-written decision procedure would reduce.

**Comment by andreasabel** [comment]
I think then it does not cover the "erase big proof terms" feature any more.
```agda
erase : {A : Set} {x y : A} (p : x == y) -> x == y
erase refl = primTrustMe
```

**Comment by UlfNorell** [comment]
I think it does. The cases where this erasure matters is when `x` and `y` are closed terms and the proof would reduce to `refl`. I think it's very rare that you erase a proof of neutral `x` and `y` that happen to be definitionally equal.

Btw, you shouldn't match on the proof in your `erase` function!

**Comment by UlfNorell** [comment]
Also, once you fix your `erase` function:
```agda
erase : {A : Set} {x y : A} (p : x == y) -> x == y
erase _ = primTrustMe
```
you'll notice that even for neutral `x` and `y` the proof term goes away.

**Comment by andreasabel** [comment]
(That's what I had originally, but then got unsure.)  
Yes, the proof term is gone, but `erase {A} {x} {x} refl` does not reduce to `refl` any more if we restrict to closed terms in `primTrustMe`.

**Comment by UlfNorell** [comment]
Indeed. I'm proposing to make `primTrustMe` a conservative extension of the theory, so naturally we'll lose some examples that the currently over-powered `primTrustMe` can handle.

I don't think losing this particular reduction is a big problem. If `refl` works for neutral terms there really shouldn't be a big proof to erase, since you would've used `refl` to begin with.

**Comment by jespercockx** [comment]
Note that you can still get (something very close to) the current behaviour of `primTrustMe` by using rewriting:
```agda
{-# OPTIONS --rewriting #-}

module _ where

open import Agda.Builtin.Equality

{-# BUILTIN REWRITE _≡_ #-}

module _ {ℓ} (A : Set ℓ) where
  postulate
    primTrustMe : {x y : A} → x ≡ y
    primTrustMe-refl : {x : A} → primTrustMe {x} {x} ≡ refl

{-# REWRITE primTrustMe-refl #-}
```
So I think it's fine to change the behaviour of `primTrustMe` to be more conservative, since there's still a way to get the current behaviour for people who really want it.

**Comment by andreasabel** [comment]
Great!  That sounds nice, the non-conservative primTrustMe is subsumed under rewriting!
We will have to implement the eta-aware equality check for the sake of rewriting then.

**Comment by andreasabel** [comment]
@UlfNorell (CC: @jespercockx): How does your proposal relate to 
- #5784

**Comment by UlfNorell** [comment]
That problem was that we only computed whnf instead of normalising. This proposal is about primEraseEquality on neutral terms and I believe the op in that case had canonical terms?

**Comment by andreasabel** [comment]
So that would mean we would reduce the terms under comparison, 
- give up when we reach neutrals 
- block when we reach blocked things
- continue comparison if we get canonicals
- apply eta when one is a neutral and one is a canonical and continue ?

Is that the idea?
